opensrp-opd/src/main/java/org/smartregister/opd/utils/OpdReverseJsonFormUtils.java

```diff
@@ -0,0 +1,160 @@
+import android.support.annotation.Nullable;
+import android.text.TextUtils;
+
+import com.google.common.reflect.TypeToken;
+
+import org.apache.commons.lang3.StringUtils;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.smartregister.domain.Photo;
+import org.smartregister.location.helper.LocationHelper;
+import org.smartregister.opd.enums.LocationHierarchy;
+import org.smartregister.opd.pojos.OpdMetadata;
+import org.smartregister.util.AssetHandler;
+import org.smartregister.util.FormUtils;
+import org.smartregister.util.ImageUtils;
+import org.smartregister.util.Utils;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+
+import timber.log.Timber;
+
+public class OpdReverseJsonFormUtils {
+
+    @Nullable
+    public static String prepareJsonEditOpdRegistrationForm(@NonNull Map<String, String> detailsMap, @NonNull List<String> nonEditableFields, @NonNull Context context) {
+        try {
+            OpdMetadata opdMetadata = OpdUtils.metadata();
+
+            if (opdMetadata != null) {
+                JSONObject form = new FormUtils(context).getFormJson(opdMetadata.getOpdRegistrationFormName());
+                Timber.d("Original Form %s", form);
+                if (form != null) {
+                    OpdJsonFormUtils.addRegLocHierarchyQuestions(form, OpdConstants.JSON_FORM_KEY.ADDRESS_WIDGET_KEY, LocationHierarchy.ENTIRE_TREE);
+                    form.put(OpdConstants.JSON_FORM_KEY.ENTITY_ID, detailsMap.get(OpdConstants.KEY.BASE_ENTITY_ID));
+
+                    form.put(OpdConstants.JSON_FORM_KEY.ENCOUNTER_TYPE, opdMetadata.getUpdateEventType());
+                    form.put(OpdJsonFormUtils.CURRENT_ZEIR_ID, Utils.getValue(detailsMap, OpdConstants.KEY.OPENSRP_ID, true).replace("-", ""));
+
+                    form.getJSONObject(OpdJsonFormUtils.STEP1).put(OpdConstants.JSON_FORM_KEY.FORM_TITLE, OpdConstants.JSON_FORM_KEY.OPD_EDIT_FORM_TITLE);
+
+                    JSONObject metadata = form.getJSONObject(OpdJsonFormUtils.METADATA);
+                    metadata.put(OpdJsonFormUtils.ENCOUNTER_LOCATION, OpdUtils.getAllSharedPreferences().fetchCurrentLocality());
+                    JSONObject stepOne = form.getJSONObject(OpdJsonFormUtils.STEP1);
+                    JSONArray jsonArray = stepOne.getJSONArray(OpdJsonFormUtils.FIELDS);
+
+                    for (int i = 0; i < jsonArray.length(); i++) {
+                        JSONObject jsonObject = jsonArray.getJSONObject(i);
+                        setFormFieldValues(detailsMap, nonEditableFields, jsonObject);
+                    }
+                    Timber.d("Final Form %s", form);
+                    return form.toString();
+                } else {
+                    Timber.e("Form cannot be found");
+                }
+            } else {
+                Timber.e(new Exception(), "Could not start OPD Edit Registration Form because OpdMetadata is null");
```

Comment:
I think we do, I need the stacktrace and exception location otherwise, only the message will be printed and I have to search for the message throughout the codebase